composer/callbacks/checkpoint_saver.py

```diff
@@ -532,8 +605,38 @@ def _save_checkpoint(self, state: State, logger: Logger):
         if self.num_checkpoints_to_keep >= 0:
             self._rotate_checkpoints(sharding_enabled=state.fsdp_sharded_state_dict_enabled)
 
-    def _rotate_checkpoints(self, sharding_enabled: bool = False):
+    def wait(self) -> None:
+        """Wait exsiting upload tasks to finish and start uploading symlink file if necessary."""
+        if self.remote_uploader is None:
+            return
+        # Wait remote uploader futures and start to upload the latest symlink file if necessary
+        if self.rank_saves_remote_symlinks and len(self.remote_uploader_futures) != len(self.symlink_file_tasks):
+            raise RuntimeError(
+                f'Expect len(remote_uploader_futures) == len(symlink_file_tasks), but got {len(self.remote_uploader_futures)} != {len(self.symlink_file_tasks)}',
+            )
+        log.debug('Waiting for previous checkpoint files upload finish')
+        for i in range(len(self.remote_uploader_futures)):
+            for future in self.remote_uploader_futures[i]:
+                future.result()
+        log.debug(f'Current rank finished existing uploading tasks')
+        self.remote_uploader_futures = []
+
+        t = dist.get_device(None).tensor_to_device(torch.tensor(1))
+        dist.all_reduce(t)
```

Comment:
Why does this not block training during uploading?

Reply:
the orginal method would block training if the uploading hasn't finished . in the latstet commit, I changed the way to upload the symlink and made it completely async.